Bluetooth: smp: Added Legacy only pairing and Legacy OOB pairing support

ivan98ams opened the pull request:
- Added public API in `conn.h` as `bt_le_oob_set_legacy_tk(u8_t *tk)`
- Added new Kconfig options:
 
1.  `CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY` to force usage of legacy OOB pairing and free up memory

1.  `CONFIG_BT_SMP_DISABLE_LEGACY_JW_PASSKEY` to forbid the usage of the insecure legacy pairing methods

Fixes: #2585

i-moraless committed:
Bluetooth: smp: Adding Legacy OOB pairing support

Added support for Legacy pairing using OOB Temporary Key

Signed-off-by: Iván Morales <ivan98ams@gmail.com>

i-moraless committed:
Bluetooth: smp: Adding Legacy OOB only mode and no JW/Passkey mode

Added CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY option that completely disables
all legacy and SC pairing modes(except for Out of Band) and frees the
memory previously used by these.

Added CONFIG_BT_SMP_DISABLE_LEGACY_JW_PASSKEY option that force rejects
pair requests that lead to legacy Just Works or Passkey pairing.

Signed-off-by: Iván Morales <ivan98ams@gmail.com>

jhedberg commented:
I'm fine with adding legacy OOB pairing as a new feature, but making it possible to disable Secure Connections support is something that requires some *really* good justification. We might as well call that option BT_MAKE_MY_PRODUCT_INSECURE. Is it even possible to qualify such a product?

jhedberg commented:
To follow up on my previous comment: the only scenario that I could consider eliminating SC code is if that would *only* permit legacy OOB pairing (which I assume can be made secure), and no other legacy method. However, I'd still urge you to check the SM ICS to verify if such a configuration is even qualifiable - I have a feeling it might not be.

ivan98ams commented:
The option to disable Secure Connections would allow to "force" the use of legacy OOB which is the only way of providing a 128bit long LL encryption key, which gives a lot more freedom to an advanced user/developer to use different/faster key agreement methods than DHKE which can also be _properly_ authenticated. This can be made "secure" think of ephemeral QR codes as an example.

I will check the document.

fnde-ot commented:
> To follow up on my previous comment: the only scenario that I could consider eliminating SC code is if that would _only_ permit legacy OOB pairing (which I assume can be made secure), and no other legacy method. However, I'd still urge you to check the SM ICS to verify if such a configuration is even qualifiable - I have a feeling it might not be.

While I agree with the principle, having the option to disable Secure Connections would allow to test that the stack works fine in legacy pairing mode (Zephyr master against Zephyr slave). I would actually prefer this to be a runtime option when calling bt_conn_set_security (it should of course default to SC).

ivan98ams commented:
> While I agree with the principle, having the option to disable Secure Connections would allow to test that the stack works fine in legacy pairing mode (Zephyr master against Zephyr slave). I would actually prefer this to be a runtime option when calling bt_conn_set_security (it should of course default to SC).

I think it would be possible to have both options, and to also include a "Legacy OOB only" config should be qualifiable configuration as it would essentially be a no input output device that only allows L3 security

joerchan commented:
> I'd still urge you to check the SM ICS to verify if such a configuration is even qualifiable - I have a feeling it might not be.

I asked our qualification guy and he says there are no restrictions when it comes to qualifying with Legacy pairing.

I understand the hesitation to disable LE SC pairing completely as these methods are really insecure.
What we have done for these pairing is to turn down the TX power to the lowest possible setting during the pairing procedure. This means that you have to hold the devices close together.

We still allow to use legacy pairing with passkey and just works. So I think if anything we could have another option to disable these two completely (with or without LE SC).

In any case I would like to get the application the options to do what they want to do. As long as we document clearly and use recommended default values.

For legacy OOB, the security is only as strong as the OOB mechanism, so we are giving away the security responsebility in this case also.

fnde-ot commented:
> I asked our qualification guy and he says there are no restrictions when it comes to qualifying with Legacy pairing.

So would it also be okay if we add (in another PR) a runtime option to bt_conn_set_security or similar to force legacy pairing? If so, we might consider implementing that.

jhedberg commented:
> While I agree with the principle, having the option to disable Secure Connections would allow to test that the stack works fine in legacy pairing mode

@fnde-ot we do have that already covered since it's something we test periodically against the PTS. Manual testing can also be done against devices that don't support SC (or BlueZ which I think allows disabling this at runtime). If testing would be the only concern then adding a new Kconfig option becomes less of an issue, since we'd make it depend on the BT_TESTING option.

fnde-ot commented:
> @fnde-ot we do have that already covered since it's something we test periodically against the PTS. Manual testing can also be done against devices that don't support SC (or BlueZ which I think allows disabling this at runtime). If testing would be the only concern then adding a new Kconfig option becomes less of an issue, since we'd make it depend on the BT_TESTING option.

We would also like to test that legacy devices cannot access some of our GATT services (only allowed for peers using SC), so a runtime option would be convenient. However, a Kconfig option is probably good enough.

joerchan commented:
@fnde-ot Runtime setting the LE SC bit I think should be a testing only-feature. If we have include the LE SC code, we should use it.

carlescufi commented:
I think that the only justification to completely disable LE SC is to force Legacy OOB. So if this option is introduced it should then only allow Legacy OOB, but no other legacy pairing method in my opinion. Would that meet @ivan98ams 's requirements?

ivan98ams commented:
@carlescufi yes, this would meet my requirements

fnde-ot commented:
> I think that the only justification to completely disable LE SC is to force Legacy OOB. So if this option is introduced it should then only allow Legacy OOB, but no other legacy pairing method in my opinion.

I'd prefer if we kept the Kconfig option to disable LE SC independent from OOB. See my above comments about our need to test our application against a Zephyr master device that pairs in legacy mode.
Maybe we could make it depend on either OOB or BT_TESTING, so to make sure noone enables it by mistake?

joerchan commented:
Ok. then I think FORCE_LEGACY_OOB is the one option that we can agree on.

Given the discussion in this PR it seems that with the introduction with Legacy OOB we might want to have an option where either do pairing with LE SC or Legacy OOB.
We already have an option called BT_SMP_SC_ONLY, which disables Legacy pairing.
Should we perhaps have an option which only disables Legacy JW or Legacy Passkey?

So my suggestion is to add two new options:
 - Force Legacy OOB only
 - Disable Legacy JW and Passkey

@fnde-ot Maybe it would be best to take testing Legacy Pairing as a separate discussion?

ivan98ams commented:
> So my suggestion is to add two new options:
> 
> * Force Legacy OOB only
> * Disable Legacy JW and Passkey

@jhedberg @carlescufi do we all agree on this?

jhedberg commented:
>> So my suggestion is to add two new options:
>> 
>> * Force Legacy OOB only
>> * Disable Legacy JW and Passkey

> @jhedberg @carlescufi do we all agree on this?

This sounds like a better proposal to me, since neither of these would directly risk degrading the security of the product. I hope we'll be able to do it with as little #ifdefs as possible, since the readability of the code is already suffering from it, meaning please look for ways of using `IS_ENABLED()` where possible.

carlescufi commented:
> > > So my suggestion is to add two new options:
> > > 
> > > * Force Legacy OOB only
> > > * Disable Legacy JW and Passkey
> 
> > @jhedberg @carlescufi do we all agree on this?
> 
> This sounds like a better proposal to me, since neither of these would directly risk degrading the security of the product. I hope we'll be able to do it with as little #ifdefs as possible, since the readability of the code is already suffering from it, meaning please look for ways of using `IS_ENABLED()` where possible.

I also agree with this proposal (which I discussed offline with @joerchan).

ivan98ams commented:
> I also agree with this proposal (which I discussed offline with @joerchan).

Great! Then I will proceed with the implementation.

sjanc commented on subsys/bluetooth/host/smp.c:
```diff
@@ -4903,6 +4913,18 @@ int bt_smp_auth_passkey_confirm(struct bt_conn *conn)
 	return 0;
 }
 
+int bt_smp_le_oob_set_tk(u8_t *tk)
```
I have mixed feeling about this kind of global switch API,  shouldn't this be configurable per remote?
Also would be nice to have const pointers where possible so that one can pass const data without casting.

ivan98ams replied:
In this case I was following the same pattern as for legacy passkey and the `bt_passkey_set()` function. Alternatively, it could be added to `struct bt_smp` similarly to how SC OOB data is set

ivan98ams replied:
@sjanc I propose then a different API:
```
int bt_le_oob_set_legacy_tk(struct bt_conn *conn, const u8_t *tk);
```

That would be backed by adding the following to `smp.c`
```
/* SMP channel specific context */
struct bt_smp {
        ...
        /* Legacy OOB tk */
        u8_t                            oob_legacy_tk[16];
        ...
}
```

joerchan replied:
@ivan98ams Sorry for the delay.

You copied the use case of `int bt_passkey_set(unsigned int passkey);` which is used for Fixed passkey. This is not what we want.

I would prefer instead that we had a callback similar to `void (*passkey_entry)(struct bt_conn *conn);` and it's matching `int bt_conn_auth_passkey_entry(struct bt_conn *conn, unsigned int passkey);`

For LE SC with OOB the similar handling is with 
`	void (*oob_data_request)(struct bt_conn *conn, struct bt_conn_oob_info *info);`
And it's expected function to be called:
In fact oob_data_request can already be used. Setting `bt_conn_oob_info.type = BT_CONN_OOB_LE_LEGACY`

bt_le_oob_set_sc_data

joerchan reviewed (changes requested):
currently being encrypted with TK=0..0

joerchan commented:
@ivan98ams I can fix it. A couple of missing pieces.

ivan98ams commented:
> @ivan98ams I can fix it. A couple of missing pieces.

But my debug prints show that the used Key is the one i am passing to `bt_le_oob_set_legacy_tk` from inside the `oob_data_request` callback
![image](https://user-images.githubusercontent.com/15169596/73674976-eac29d80-46b1-11ea-9e67-fc2ccf267326.png)

joerchan committed:
Bluetooth: shell: Add support for Legacy OOB pairing in the shell

Add command to input the legacy pairing OOB TK during the pairing
procedure.
Fixed shell build with CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY defined.
Print error code when passkey input failed.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1109,6 +1109,10 @@ int bt_conn_set_security(struct bt_conn *conn, bt_security_t sec)
+	if (IS_ENABLED(CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY) &&
+	    sec > BT_SECURITY_L3)
+		return -EOPNOTSUPP;
+
```
Missing { }

ivan98ams replied:
checkpatch complains about style because it is a one line if statement

jhedberg replied:
I believe you had other one-liners with {} in this same PR? Anyway, you might get complaints if you run checkpatch yourself without disabling this particular check. The Zephyr coding style however *requires* {} as you'll probably see if you look at any existing Zephyr code. The Zephyr CI checks have this checkpatch feature disabled so you shouldn't get complaints for the PR.

joerchan replied:
I added those {}, and ran checkpatch, I didn't see any warnings from checkpatch.

ivan98ams replied:
Aha, nvm then will readd

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -374,6 +390,8 @@ static u8_t get_pair_method(struct bt_smp *smp, u8_t remote_io)
 
 	return gen_method_sc[remote_io][get_io_capa()];
+#endif
+	return JUST_WORKS;
```
Shouldn't there be a `#else` branch here? Otherwise you'll generate code that has two consecutive unconditional return statements?

ivan98ams replied:
Agree

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -684,6 +702,17 @@ static bool update_keys_check(struct bt_smp *smp)
 
+	if (IS_ENABLED(CONFIG_BT_SMP_DISABLE_LEGACY_JW_PASSKEY) &&
+	    !atomic_test_bit(smp->flags, SMP_FLAG_SC) &&
+	    (smp->method != LEGACY_OOB)) {
```
minor: you can drop the ()

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -684,6 +702,17 @@ static bool update_keys_check(struct bt_smp *smp)
+	}
+
+	if (IS_ENABLED(CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY) &&
+	    (smp->method != LEGACY_OOB)) {
```
minor: you can drop the ()

jhedberg reviewed (changes requested):
A couple of minor things I spotted (after a very quick read-through though)

ivan98ams commented:
@jhedberg suggested changes were applied

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -375,6 +390,9 @@ static u8_t get_pair_method(struct bt_smp *smp, u8_t remote_io)
 	}
 
 	return gen_method_sc[remote_io][get_io_capa()];
+#elif
```
Don't you mean `#else` here? `#elif` requires a condition AFAIK

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -4925,6 +4960,7 @@ int bt_smp_auth_passkey_confirm(struct bt_conn *conn)
 	}
 
 	return 0;
+
```
This addition seems unrelated (and unnecessary)

ivan98ams commented:
Changes addressed, and waiting on shippable success, any other changes @jhedberg ?
